modules/caddytls/acmeissuer.go

```diff
@@ -436,6 +447,50 @@ func onDemandAskRequest(ask string, name string) error {
 	return nil
 }
 
+func ParseCaddyfilePreferredChainsOptions(d *caddyfile.Dispenser) (*ChainPreference, error) {
+	chainPref := new(ChainPreference)
+	if d.NextArg() {
+		smallestOpt := d.Val()
+		if smallestOpt == "smallest" {
+			trueBool := true
+			chainPref.Smallest = &trueBool
+			if d.NextBlock(1) {
```

Comment:
I used to use `if d.Next() {`, but it didn't work when calling this function through the ACME issuer config. Don't know why because it worked when this wasn't wrapped in a function: https://github.com/caddyserver/caddy/blob/ecc798af4c6b3370d2a7c5442e62afab81e03aba/modules/caddytls/acmeissuer.go#L407-L441

---

`if d.NextBlock(1) {` seems to work fine both when parsing the global option and the ACME issuer option. I'm a bit confused, to be honest...

Reply:
Note [the proper way to use `NextBlock()`](https://pkg.go.dev/github.com/caddyserver/caddy/v2@v2.4.1/caddyconfig/caddyfile#Dispenser.NextBlock) is:

```
for nesting := d.Nesting(); d.NextBlock(nesting); {
}
```

especially for when you don't know the current level of nesting. Give that a try.

Reply:
I think you can do `d.NextBlock(d.Nesting())` instead, removes the hardcoded number.

`NextBlock` essentially reads the next subdirective after the next `{` if any, until the next `}` is reached. It's usually used in loops like `for nesting := h.Nesting(); h.NextBlock(nesting); {`

Reply:
Francis beat me to it. That's what I meant to say :smile:

Reply:
Ah sorry, I think I misread what you're trying to do.

I think what you wanted was actually `d.NextArg()` to check that no other argument follows `smallest`. `d.Next()` will actually read the `{` which follows which is incorrect.

Reply:
In this case I want to check that when the user sets the `smallest` option, setting `root_common_name` or `any_common_name` **shouldn't** be possible (if I've understood correctly, these options can't be combined). Using `d.NextBlock(d.Nesting())` as you suggested seems to work great for this.

I could probably add another check with `d.NextArg()` to make sure no more arguments follow `smallest`. And I guess I should also add a check to make sure that the user has configured _something_, i.e. the following config should be invalid:

```
{
  preferred_chains
}
```

Reply:
Perfect. Those ideas also sound good. There's a ton of places we don't do checks like those, but they add robustness to the parser and help prevent user errors ("I typed this and didn't get an error, but it's not working" kind of things), so they're a good thing if you think of them. :+1:

Reply:
Done and pushed!